src/txn/txn_rollback_to_stable.c

```diff
@@ -116,23 +112,28 @@ __txn_abort_newer_update(WT_SESSION_IMPL *session,
 		 * strict timestamp checking, assert that all more recent
 		 * updates were also rolled back.
 		 */
-		if (upd->txnid == WT_TXN_ABORTED && upd == first_upd)
-			first_upd = upd->next;
-		else if (__wt_timestamp_iszero(&upd->timestamp)) {
-			if (skip_zero_timestamps && upd == first_upd)
+		if (upd->txnid == WT_TXN_ABORTED ||
+		    __wt_timestamp_iszero(&upd->timestamp)) {
+			if (upd == first_upd)
 				first_upd = upd->next;
 		} else if (__wt_timestamp_cmp(
 		    rollback_timestamp, &upd->timestamp) < 0) {
-			upd->txnid = WT_TXN_ABORTED;
-			WT_STAT_CONN_INCR(session, txn_rollback_upd_aborted);
-			__wt_timestamp_set_zero(&upd->timestamp);
-
 			/*
 			 * If any updates are aborted, all newer updates
 			 * better be aborted as well.
+			 *
+			 * Validate timestamp ordering only if configured.
 			 */
-			WT_ASSERT(session, upd == first_upd);
+			WT_ASSERT(session,
+			    !FLD_ISSET(S2BT(session)->assert_flags,
+			    WT_ASSERT_COMMIT_TS_ALWAYS |
+			    WT_ASSERT_COMMIT_TS_KEYS) ||
```

Comment:
After looking more closely at `__txn_commit_timestamp_validate`, specifically https://github.com/wiredtiger/wiredtiger/blob/develop/src/txn/txn.c#L611-L615, I think we should check for `WT_ASSERT_COMMIT_TS_KEYS` here.

Reply:
Agreed. I had this in my head but was not sure.